compiler/src/iree/compiler/Codegen/Common/TopkLowering.cpp

```diff
@@ -0,0 +1,915 @@
+          newOperands.push_back(insertElemIfNeeded.getResult(3));
+          rewriter.create<scf::YieldOp>(loc, newOperands);
+          forYield.erase();
+        }
+        SmallVector<Value> operands;
+        operands.push_back(maskProcessingLoop.getResult(0));
+        operands.push_back(maskProcessingLoop.getResult(1));
+        operands.push_back(maskProcessingLoop.getResult(2));
+        operands.push_back(maskProcessingLoop.getResult(3));
+        b.create<scf::YieldOp>(loc, operands);
+      },
+      [&](OpBuilder &b, Location loc) {
+        // else
+        // No bigger elements. No work to do.
+        SmallVector<Value> operands;
+        operands.push_back(ifInitOut.getResult(0));
+        operands.push_back(ifInitOut.getResult(1));
+        operands.push_back(ifInitOut.getResult(2));
+        operands.push_back(ifInitOut.getResult(3));
+        b.create<scf::YieldOp>(loc, operands);
+      });
+
+  // If the types were not expected, don't do rewriting.
+  if (!expectedRanksAndDims)
+    return failure();
+
+  // Create a new yield op for the top (tiling) loop.
+  {
+    OpBuilder::InsertionGuard g(rewriter);
+    scf::YieldOp yieldOp =
+        llvm::cast<scf::YieldOp>(newLoop.getBody()->getTerminator());
+    rewriter.setInsertionPoint(yieldOp);
+    SmallVector<Value> operands;
+    operands.push_back(ifVecCmp.getResult(0));
+    operands.push_back(ifVecCmp.getResult(1));
+    // After first run the out arrays are initialized.
+    operands.push_back(
+        rewriter.create<arith::ConstantIntOp>(yieldOp.getLoc(), 0, 1));
+    operands.push_back(ifVecCmp.getResult(2));
+    operands.push_back(ifVecCmp.getResult(3));
+    rewriter.create<scf::YieldOp>(yieldOp.getLoc(), operands);
+    rewriter.eraseOp(yieldOp);
+  }
+  newResults.push_back(topkOp.outputValues());
+  newResults.push_back(topkOp.outputIndices());
+  return success();
+}
+
+void TopkLoweringPass::runOnOperation() {
+  MLIRContext *context = &getContext();
+  auto funcOp = getOperation();
+  IRRewriter rewriter(context);
+  SmallVector<Operation *> candidates;
+  funcOp.walk([&](Operation *op) {
+    if (isa<IREE::LinalgExt::TopkOp>(op) &&
+        isTopkSCFLowerEnabled(dyn_cast<IREE::LinalgExt::TopkOp>(op))) {
+      candidates.push_back(op);
+    }
+  });
+
+  // The vector input sizes inference needs to use producers, so we apply
+  // vectorization from bottom to top.
+  std::reverse(candidates.begin(), candidates.end());
+  for (Operation *op : candidates) {
+    if (auto topkOp = dyn_cast<IREE::LinalgExt::TopkOp>(op)) {
+      auto arg0 = topkOp.getInputs()[0];
+      auto shapedTy = llvm::cast<ShapedType>(arg0.getType());
+      SmallVector<int64_t> vectorSizes;
+      vectorSizes.append(shapedTy.getShape().begin(),
+                         shapedTy.getShape().end());
+      if (shapedTy.isDynamicShape(vectorSizes))
+        continue;
+      if (failed(lowerTopkOpPreconditionUsingSCF(topkOp))) {
```

Comment:
Maybe better to check for all precondition during candidate selection and fold all the logic into this method.

Reply:
Done.